CONTRIBUTING.md

```diff
@@ -0,0 +1,131 @@
+example) referring to internal systems, please make sure to include all of the
+essential information in the public descriptions - describe the symptoms of the
+issue, how it was detected, investigated, how it could be reproduced, what are
+the trade-offs and so on as appropriate.
+
+## Split into commits
+
+Following from the rules described above, if what the commit achieves is
+difficult to fit into its subject, it is probably better to split it into
+several commits, if possible. Note that every commit should build (`make`
+should work and the CI should pass) independently, without requiring future
+commits. This means some modifications can't really be split into several
+commits (datamodel changes, in particular, require modifications to several
+components at the same time), but makes it easier to revert part of the Pull
+Request if some issues are detected in integration testing at a later point.
+
+## Good Commit Messages
+
+Commit messages (and the body of a Pull Request) should be as helpful and
+descriptive as possible. If applicable, please include a description of current
+behaviour, your changes, and the new behaviour. Justify the reasoning behind
+your changes - are they sufficient on their own, or preparing for more changes?
+Link any appropriate documentation, issues, or commits.
+
+## CI
+
+Please make sure your Pull Request passes the Github CI. It will verify that
+your code has been properly formatted (can be done locally with `make format`),
+builds (`make` and `make check`), and passes the unit tests (`make test`).
+The CI will run in the branches of your fork, so you can verify it passes
+there before opening a Pull Request.
+
+## Testing
+
+Describe what kind of testing your contribution underwent. If the testing was
+manual, please describe the commands or external clients that were used. If the
+tests were automated, include at least a cursory description/name of the tests,
+when they were regressed, if possible.
+
+Please note that any contribution to the code of the project will likely
+require at least some testing to be done. Depending on how central the
+component touched in your PR is to the system, the more things could only be
+detected in real-world usecases through integration testing.
+
+If a commit has been determined to break integration testing at a later stage,
+please note that the first and safest measure will almost always be reverting
+the faulty commit. Making sure critical tests are passing remains a priority
+over waiting for some commit to be reworked or refactored (which can be worked
+on after a revert has been done). Though we are striving to make more tests
+public (with failure then being visible to all), as long as some critical tests
+remain private, this will also apply to such tests (with maintainers flagging
+the breakage preferrably describing at least the gist of the test).
+
+If you are still waiting on some testing to be done, please mark the PR as a
```

Comment:
If wider testing is needed (e.g. the change itself is believed to be correct but may expose latent bugs in other components), lightweight feature flags can also be used. E.g. an entry in xapi_globs.ml and xapi.conf, where the feature/change is default off, to be turned on at a future time (when e.g. more related PRs land, or it has passed some wider testing)